src/compiler/checker.ts

```diff
@@ -4092,7 +4106,15 @@ namespace ts {
                         const resolvedSymbol = resolveName(param, paramSymbol.name, SymbolFlags.Value, undefined, undefined);
                         paramSymbol = resolvedSymbol;
                     }
-                    parameters.push(paramSymbol);
+                    if (paramSymbol.name === "this") {
+                        thisType = param.type && getTypeOfSymbol(paramSymbol);
+                        if (i !== 0 || declaration.kind === SyntaxKind.Constructor) {
+                            error(param, Diagnostics.this_cannot_be_referenced_in_current_location);

```

Comment:
I would have one error message that says `'this' parameter must be the first parameter` and another that says `A constructor cannot have a 'this' parameter`.